tools/build_api.py

```diff
@@ -426,20 +430,27 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
             notify.info("  Filling region %s with %s" % (region.name, region.filename))
             part = intelhex_offset(region.filename, offset=region.start)
             part.start_addr = None
-            part_size = (part.maxaddr() - part.minaddr()) + 1
+
+            # make same assumption as in region builder; first segment must fit.
+            # this is only to get a neat ToolException anyway, since IntelHex.merge will
+            # throw IntelHex.AddressOverlapError if there's overlapping
+            part_size = 0
+            for start, stop in part.segments():
+                part_size += stop - start
+            merged.merge(part)
+
             if part_size > region.size:
                 raise ToolException("Contents of region %s does not fit"
                                     % region.name)
-            merged.merge(part)
-            pad_size = region.size - part_size
-            if pad_size > 0 and region != region_list[-1]:
-                notify.info("  Padding region %s with 0x%x bytes" %
-                            (region.name, pad_size))
-                if format is ".hex":
-                    """The offset will be in the hex file generated when we're done,
-                    so we can skip padding here"""
-                else:
-                    merged.puts(merged.maxaddr() + 1, padding * pad_size)
+
+    # Hex file can have gaps, so no padding needed. While other formats may
+    # need padding. Iterate through segments and pad the gaps.
+    if format != ".hex":
+        begin = 0
```

Comment:
Probably mean start of the first region. This will break unmanaged bootloder mode, and managed bootloder mode when the first address of rom != 0.

Reply:
Uploaded https://github.com/ARMmbed/mbed-os/pull/8097/commits/38ce7b13e7bbc7bf759d0eca76cd0d4d6076ce62